Library/Homebrew/test/utils/git_spec.rb

```diff
@@ -0,0 +1,175 @@
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      expect(described_class.git_available?).to be_falsey
+    end
+  end
+
+  describe "::git_path" do
+    it "returns nil when git is not available" do
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      expect(described_class.git_path).to eq(nil)
+    end
+
+    it "returns path of git when git is available" do
+      expect(described_class.git_path).to end_with("git")
+    end
+  end
+
+  describe "::git_version" do
+    it "returns nil when git is not available" do
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      expect(described_class.git_path).to eq(nil)
+    end
+
+    it "returns version of git when git is available" do
+      expect(described_class.git_version).not_to be_nil
+    end
+  end
+
+  describe "::ensure_git_installed!" do
+    it "returns nil if git already available" do
+      expect(described_class.ensure_git_installed!).to be_nil
+    end
+
+    context "when git is not already available" do
+      before do
+        stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      end
+
+      it "can't install brewed git if homebrew/core is unavailable" do
+        allow_any_instance_of(Pathname).to receive(:directory?).and_return(false)
+        expect { described_class.ensure_git_installed! }.to raise_error("Git is unavailable")
+      end
+
+      it "raises error if can't install git" do
+        stub_const("HOMEBREW_BREW_FILE", mktmpdir/"brew")
+        expect { described_class.ensure_git_installed! }.to raise_error("Git is unavailable")
+      end
+
+      it "installs git" do
+        allow(Homebrew).to receive(:_system).with(any_args).and_return(true)
+        described_class.ensure_git_installed!
+      end
+    end
+  end
+
+  describe "::git_remote_exists" do
+    it "returns true when git is not available" do
+      stub_const("HOMEBREW_SHIMS_PATH", mktmpdir/"shim")
+      expect(described_class.git_remote_exists("blah")).to be_truthy
+    end
+
+    context "when git is available" do
+      it "returns true when git remote exists", :needs_network do
+        git = HOMEBREW_SHIMS_PATH/"scm/git"
+        repo = Pathname.new("hey")
+        repo.mkpath
+
+        system "cd", repo
+        system git, "init"
+        system git, "remote", "add", "origin", "http://github.com/Homebrew/homebrew.github.io"
+        system "cd .."
+
+        expect(described_class.git_remote_exists("http://github.com/Homebrew/homebrew.github.io")).to be_truthy
+
+        FileUtils.rm_rf(repo)
+      end
+
+      it "returns false when git remote does not exist" do
+        expect(described_class.git_remote_exists("blah")).to be_falsey
+      end
+    end
+  end
+
+  describe "::clear_git_available_cache" do
+    it "removes @git_path and @git_version if defined" do
+      described_class.clear_git_available_cache
+
+      expect(described_class.instance_variable_get(:@git_path)).to be_nil
```

Comment:
Need to try and figure out a way to not set/get instance variables here. If need be: delete the relevant tests that need them set given we're almost at the end of GSoC. Ideally: just call other methods that end up calling these ones.